contracts/StakeManager.sol

```diff
@@ -0,0 +1,165 @@
+pragma solidity ^0.5.16;
+pragma experimental ABIEncoderV2;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+import "./interfaces/IRelayHub.sol";
+
+contract StakeManager {
+
+    /// Emitted when a stake or unstakeDelay are initialized or increased
+    event StakeAdded(
+        address indexed registree,
+        address indexed owner,
+        uint256 stake,
+        uint256 unstakeDelay
+    );
+
+    /// Emitted once a stake is scheduled for withdrawal
+    event StakeUnlocked(
+        address indexed registree,
+        address indexed owner,
+        uint256 withdrawBlock
+    );
+
+    /// Emitted when owner withdraws registree funds
+    event StakeWithdrawn(
+        address indexed registree,
+        address indexed owner,
+        uint256 amount
+    );
+
+    /// Emitted when an authorized Relay Hub penalizes a registree
+    event StakePenalized(
+        address indexed registree,
+        address indexed beneficiary,
+        uint256 reward
+    );
+
+    event HubAuthorized(
+        address indexed registree,
+        address indexed relayHub
+    );
+
+    event HubUnauthorized(
+        address indexed registree,
+        address indexed relayHub,
+        uint256 removalBlock
+    );
+
+    /// @param stake - amount of ether staked for this relay
+    /// @param unstakeDelay - number of blocks to elapse before the owner can retrieve the stake after calling 'unlock'
+    /// @param withdrawBlock - first block number 'withdraw' will be callable, or zero if the unlock has not been called
+    /// @param owner - address that receives revenue and manages registree's stake
+    struct StakeInfo {
+        uint256 stake;
+        uint256 unstakeDelay;
+        uint256 withdrawBlock;
+        address payable owner;
+    }
+
+    struct RelayHubInfo {
+        uint256 removalBlock;
+    }
+
+    /// maps owners to their registrees
+    mapping(address => address) public registrees;
+
+    /// maps registrees to their stakes
+    mapping(address => StakeInfo) public stakes;
+
+    /// maps registrees to their authorized hubs map
+    mapping(address => mapping(address => RelayHubInfo)) public authorizedHubs;
+
+    /// Put a stake for a registree and set its unstake delay.
+    /// If the entry does not exist, it is created, and the caller of this function becomes its owner.
+    /// If the entry already exists, only the owner can call this function.
+    /// @param registree - address that represents a stake entry and controls relay registrations on relay hubs
+    /// @param unstakeDelay - number of blocks to elapse before the owner can retrieve the stake after calling 'unlock'
+    function stakeForAddress(address registree, uint256 unstakeDelay) external payable {
+        require(registrees[msg.sender] == address(0) || registrees[msg.sender] == registree, "different stake exists");
```

Comment:
at least the comment string is wrong (no stake is defined yet)

Reply:
This particular require? This means this owner already made a stake earlier, and for a different registree address. I don't support it, I think you'll agree this is too much complexity